### PR TITLE
Flutter/tests

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: scanners/flutter/flutter.go [hunk 1]
```diff
@@ -152,40 +157,62 @@ func (scanner *Scanner) DetectPlatform(searchDir string) (bool, error) {
 			return false, err
 		}
 
+		testsDirPath := filepath.Join(projectLocation, "test")
+		if exists, err := pathutil.IsDirExists(testsDirPath); err == nil && exists {
+			if files, err := ioutil.ReadDir(testsDirPath); err == nil && len(files) > 0 {
+				for _, file := range files {
+					if strings.HasSuffix(file.Name(), "_test.dart") {
+						proj.hasTest = true
```
Comment (godrei) [inline scanners/flutter/flutter.go]: break, as it is enough to set it once

File: scanners/flutter/flutter.go [hunk 2]
```diff
@@ -214,58 +241,130 @@ func (scanner *Scanner) Options() (models.OptionModel, models.Warnings, error) {
...
-
-			for xcodeWorkspacePath, schemes := range project.xcodeProjectPaths {
-				schemeOption := models.NewOption(ios.SchemeInputTitle, ios.SchemeInputEnvKey)
-				projectPathOption.AddOption(xcodeWorkspacePath, schemeOption)
-
-				for _, scheme := range schemes {
-					exportMethodOption := models.NewOption(ios.IosExportMethodInputTitle, ios.ExportMethodInputEnvKey)
-					schemeOption.AddOption(scheme, exportMethodOption)
+		if project.hasTest {
+			flutterProjectHasTestOption := models.NewOption(testsInputTitle, "_")
```
Comment (godrei) [inline scanners/flutter/flutter.go]: make sure an env with key `_` is not added to the `app/envs` in the generated bitrise.yml

File: scanners/flutter/flutter.go [hunk 3]
```diff
@@ -214,58 +241,130 @@ func (scanner *Scanner) Options() (models.OptionModel, models.Warnings, error) {
...
+									configOption := models.NewConfigOption(cfg + "-app-" + getBuildablePlatform(project.hasAndroidProject, project.hasIosProject))
+									exportMethodOption.AddConfig(exportMethod, configOption)
+								}
+							}
+						}
+					} else {
+						configOption := models.NewConfigOption(cfg + "-app-" + getBuildablePlatform(project.hasAndroidProject, project.hasIosProject))
+						flutterProjectHasTestOption.AddOption(v, configOption)
 					}
+				} else {
```
Comment (godrei) [inline scanners/flutter/flutter.go]: is a flutter project valid without ios or android parts?

File: scanners/flutter/flutter.go [hunk 4]
```diff
@@ -214,58 +241,130 @@ func (scanner *Scanner) Options() (models.OptionModel, models.Warnings, error) {
...
+								schemeOption.AddOption(scheme, exportMethodOption)
+
+								for _, exportMethod := range ios.IosExportMethods {
+									configOption := models.NewConfigOption(cfg + "-app-" + getBuildablePlatform(project.hasAndroidProject, project.hasIosProject))
+									exportMethodOption.AddConfig(exportMethod, configOption)
+								}
+							}
+						}
+					} else {
+						configOption := models.NewConfigOption(cfg + "-app-" + getBuildablePlatform(project.hasAndroidProject, project.hasIosProject))
```
Comment (godrei) [inline scanners/flutter/flutter.go]: do not have any only android specific question?

File: scanners/flutter/flutter.go [hunk 5]
```diff
@@ -279,69 +378,93 @@ func (scanner *Scanner) Configs() (models.BitriseConfigMap, error) {
...
+	for _, variant := range []struct {
+		configID string
+		test     bool
+		deploy   bool
+		platform string
+	}{
+		{test: false, deploy: false, configID: configName},
+		{test: true, deploy: false, configID: configName + "-test"},
+		{test: false, deploy: true, platform: "both", configID: configName + "-app-both"},
+		{test: true, deploy: true, platform: "both", configID: configName + "-test-app-both"},
```
Comment (godrei) [inline scanners/flutter/flutter.go]: no test, no deploy is not valid for platforms: both?

File: _tests/integration/flutter_test.go [hunk 6]
```diff
@@ -74,20 +74,1118 @@ func TestFlutter(t *testing.T) {
...
+	steps.DeployToBitriseIoVersion,
+}
+
+var flutterSampleAppResultYML = fmt.Sprintf(`options:
+  flutter:
+    title: Project Location
+    env_key: BITRISE_FLUTTER_PROJECT_LOCATION
+    value_map:
+      .:
+        title: Run tests found in the project
```
Comment (godrei) [inline _tests/integration/flutter_test.go]: missing `?` mark?

File: scanners/flutter/flutter.go [hunk 7]
```diff
@@ -32,6 +39,12 @@ var (
 		"plugin",
 		"package",
 	}
+	platforms = []string{
+		"none",
```
[the remark this answers is not in the document]
Comment (godrei) [reply scanners/flutter/flutter.go]: discussed

File: scanners/flutter/flutter.go (hunk 3 above)
Comment (godrei) [reply scanners/flutter/flutter.go]: discussed

File: scanners/flutter/flutter.go (hunk 4 above)
Comment (godrei) [reply scanners/flutter/flutter.go]: discussed

File: scanners/flutter/flutter.go (hunk 5 above)
Comment (godrei) [reply scanners/flutter/flutter.go]: discussed

File: _tests/integration/flutter_test.go (hunk 6 above)
Comment (godrei) [reply _tests/integration/flutter_test.go]: discussed

File: _tests/integration/flutter_test.go [hunk 8]
```diff
@@ -74,20 +74,1118 @@ func TestFlutter(t *testing.T) {
...
+        env_key: _
```
[the remark this answers is not in the document]
Comment (trapacska) [reply _tests/integration/flutter_test.go]: website is using the only env keys we set in the yml app envs section.

File: scanners/flutter/flutter.go (hunk 2 above)
Comment (trapacska) [reply scanners/flutter/flutter.go]: website is using the only env keys we set in the yml app envs section.

screen shot 2019-01-21 at 3 12 54 pm:
screen shot 2019-01-21 at 3 12 30 pm

with `_`:
screen shot 2019-01-21 at 3 13 33 pm:
screen shot 2019-01-21 at 3 13 26 pm

**Pull request merged**